### ISSUE TITLE
[Postponed] Write a GHC plugin that transforms pure functions to monadic functions with delta-let

### ISSUE THREAD

**Issue by Mikolaj** [Mikolaj/horde-ad#6]
As a potential solution to #64:

Simon proposed to free the user from the tedium of inserting delta-lets in the proper places and, more generally, of working in a monad, by employing a GHC plugin for this purpose.

My impression from benchmarks is that indeed inserting delta-lets everywhere, as in the paper, causes a very mild slowdown (30% is extreme cases, such as benchmark `prod`, but normally a couple percent). However, the plugin could spot easy cases of unnecessary delta-lets (variables used immediately, only once and without duplication) and so probably eliminate the slowdown in practice. In this way, the only benefit of user controlling delta-lets is gone, together with the risk of missing a required delta-let and causing deta-expression explosion.

[Edit: this is no longer the case: the monadic code can now be rank-polymorphic and also @tomjaguarpaw rightly notes this kind of polymorphism is not essential, even though @Mikolaj likes the marginal code de-duplication it enables.] _Another advantage of the plugin is that from one definition it would produce monadic functions for all primitive scalar and non-scalar differentiable types (such as vector or matrices). So far, monadic functions can't be polymorphic wrt scalar/vector/matrix, so without a plugin, there is a bit of code repetition._ 

The plugin could also produce a version of the function bereft of any gradient machinery and so well suited for testing the trained parameters or using them for the actual application they are trained for. For now, a special instance of our special monad class is used for computing the normal value of a function, many times faster than the normal monad instance producing a gradient, but a couple of times slower than a bare non-monadic function would be. [Edit: in the new code, the special instance is even closer to the original function in performance, but it may never completely reach it and it would depend on GHC being exceptionally smart, anyway, which is naturally fragile.]

Due to these considerations, I guess it's better to wait with designing the plugin until we settle on the set of primitive differentiable types and the details of our monad class and its instances specialized for various tasks.

**Comment by Mikolaj** [comment]
BTW, the type-level plugins we depend on are close to being ported to GHC 9.4 (and so probably to HEAD, as well): https://github.com/clash-lang/ghc-tcplugins-extra/pull/22. Once this happens we can start writing our plugin using the new GHC plugin API and so have it future-proof. Also, GHC HEAD fixes (or will fix soon) our performance problems, so we'd rather code for GHC 9.6 or later.

**Comment by Mikolaj** [comment]
For the TH approach, here's an interesting piece from a competitor: https://tomsmeding.com/f/ad-dualrev-th-docs/Language-Haskell-ReverseAD-TH.html (https://github.com/tomsmeding/ad-dualrev-th).

**Comment by Mikolaj** [comment]
A similary simple plugin, I think: https://brandonchinn178.github.io/blog/2022/07/11/writing-a-ghc-plugin-and-preprocessor-for-the-first-time.html

> I decided to follow common practice and only support the last three GHC versions (8.10, 9.0, 9.2), but it was still a decent amount of effort. Here are a couple examples of the breaking changes in these three versions...

I hope by supporting only 9.4, at least initially, we can avoid the CPP noise, etc.

**Comment by Mikolaj** [comment]
Could somebody, please, elaborate what the requirements for this ticket are? I'm beginning to think a plugin or TH solution would have some of the problems my impure solution has plus some of the problems of the explicitly monadic code. This also has a relation to the restriction in the paper to differentiating closed expressions, while in practice expressions are not closed, because users use their common toolbox of definitions and also Haskell floats some expressions to top level and so they end up getting used in many gradient computation runs. This may lead, e.g., to mixup of variable (or Tree0 node) identities between runs (assuming the runs assign the (not yet assigned?) identities independently).

The be concrete: let's say I have top level expression e, used in top level expressions e1 and e2. I want to compute gradients of e, e1, e2 in the same run of my executable. (This is a practical concern, because normally I want to compute hundreds of gradients in a gradient descent loop and the gradients are not always of the same expression.) What is the user expected to write and what the plugin and TH are supposed to do?

```
e = 0
e1 = e + e + 1
e2 = e + 2

main = do
  let gradient_e = ...
      gradient_1 = ...
      gradient_2 = ...
 print (gradient_e, gradient_e1, gradient_e2)
```

**Comment by Mikolaj** [comment]
Let me close this until we decide to revert the "impure Tree0" solution (#62) that accidentally solves the same problem (#64) without a plugin or TH. There are a few details needed to close #64, but plugins should not be necessary to tackle them or at least not worth the trouble.
